Extension/src/LanguageServer/cppBuildTaskProvider.ts

```diff
@@ -0,0 +1,342 @@
+        if (selectedTask) {
+            return;
+        }
+
+        const buildTasks: vscode.Task[] = await this.getTasks(true);
+        selectedTask = buildTasks.find(task => task.name === taskName);
+        console.assert(selectedTask);
+        if (!selectedTask) {
+            throw new Error("Failed to get selectedTask in ensureBuildTaskExists()");
+        }
+
+        rawTasksJson.version = "2.0.0";
+
+        const selectedTask2: vscode.Task = selectedTask;
+        if (!rawTasksJson.tasks.find((task: any) => task.label === selectedTask2.definition.label)) {
+            const task: any = {
+                ...selectedTask2.definition,
+                problemMatcher: selectedTask2.problemMatchers,
+                group: { kind: "build", "isDefault": true }
+            };
+            rawTasksJson.tasks.push(task);
+        }
+
+        // TODO: It's dangerous to overwrite this file. We could be wiping out comments.
+        const settings: OtherSettings = new OtherSettings();
+        const tasksJsonPath: string | undefined = getTasksJsonPath();
+        if (!tasksJsonPath) {
+            throw new Error("Failed to get tasksJsonPath in ensureBuildTaskExists()");
+        }
+
+        await util.writeFileText(tasksJsonPath, JSON.stringify(rawTasksJson, null, settings.editorTabSize));
+    }
+}
+
+class CustomBuildTaskTerminal implements vscode.Pseudoterminal {
+    private writeEmitter = new vscode.EventEmitter<string>();
+    private closeEmitter = new vscode.EventEmitter<number>();
+    public get onDidWrite(): vscode.Event<string> { return this.writeEmitter.event; }
+    public get onDidClose(): vscode.Event<number> { return this.closeEmitter.event; }
+
+    constructor(private command: string, private args: string[], private options: cp.ExecOptions | undefined) {
+    }
+
+    async open(_initialDimensions: vscode.TerminalDimensions | undefined): Promise<void> {
+        telemetry.logLanguageServerEvent("cppBuildTaskStarted");
+        // At this point we can start using the terminal.
+        this.writeEmitter.fire("Starting build...\r\n");
+        await this.doBuild();
+    }
+
+    close(): void {
+        // The terminal has been closed. Shutdown the build.
+    }
+
+    private async doBuild(): Promise<any> {
+        // Do build.
+        let activeCommand: string = util.resolveVariables(this.command, this.AdditionalEnvironment);
+        this.args.forEach(value => {
+            let temp: string = util.resolveVariables(value, this.AdditionalEnvironment);
+            if (temp && temp.includes(" ")) {
+                temp = "\"" + temp + "\"";
+            }
+            activeCommand = activeCommand + " " + temp;
+        });
+        if (this.options?.cwd) {
+            this.options.cwd = util.resolveVariables(this.options.cwd, this.AdditionalEnvironment);
+        }
+        await new Promise<number>((resolve, reject) => {
```

Comment:
If you await something that will potentially `reject` the promise, you should wrap it with `try/finally`.  Though it probably makes more sense to resolve to -1 than to reject the promise.  Give this a try:
```
        try {
            const result = await new Promise<number>((resolve, reject) => {
                cp.exec(activeCommand, this.options, (_error, stdout, _stderr) => {
                    if (_error) {
                        telemetry.logLanguageServerEvent("cppBuildTaskError");
                        this.writeEmitter.fire("Build finished with error:\r\n");
                        this.writeEmitter.fire(stdout.toString());
                        resolve(-1);
                    } else {
                        this.writeEmitter.fire(stdout.toString());
                        this.writeEmitter.fire("\r\nBuild finished successfully.\r\n");
                        resolve(0);
                    }
                });
            });
            this.closeEmitter.fire(result);
        } catch {
            this.closeEmitter.fire(-1);
        }
```

Reply:
```
        try {
            await new Promise<number>((resolve, reject) => {
                cp.exec(activeCommand, this.options, (_error, stdout, _stderr) => {
                    if (_error) {
                        telemetry.logLanguageServerEvent("cppBuildTaskError");
                        const endOfLine: string = stdout ? ":" : ".";
                        this.writeEmitter.fire("Build finished with error" + endOfLine + "\r\n");
                        this.writeEmitter.fire(stdout.toString());
                        resolve(-1);
                    } else {
                        this.writeEmitter.fire(stdout.toString());
                        this.writeEmitter.fire("\r\nBuild finished successfully.\r\n");
                        resolve(0);
                    }
                });
            });
            this.closeEmitter.fire(0);
        } catch {
            this.closeEmitter.fire(-1);
        }
```